v2/svid/x509svid/svid.go

```diff
@@ -136,41 +118,165 @@ func (s *SVID) GetX509SVID() (*SVID, error) {
 	return s, nil
 }
 
-func newSVID(certificates []*x509.Certificate, privateKey interface{}) (*SVID, error) {
+func newSVID(certificates []*x509.Certificate, privateKey crypto.PrivateKey) (*SVID, error) {
+	spiffeID, err := validateCertificates(certificates)
+	if err != nil {
+		return nil, x509svidErr.New("certificate validation failed: %v", err)
+	}
+
+	signer, err := validatePrivateKey(privateKey, certificates[0])
+	if err != nil {
+		return nil, x509svidErr.New("private key validation failed: %v", err)
+	}
+
+	return &SVID{
+		Certificates: certificates,
+		PrivateKey:   signer,
+		ID:           *spiffeID,
+	}, nil
+}
+
+// validate the slice of certificates constitutes a valid SVID chain according
+// to the spiffe standard and returns the spiffe id of the leaf certificate
+func validateCertificates(certificates []*x509.Certificate) (*spiffeid.ID, error) {
```

Comment:
This is a very thorough validation of the SVID certificate chain. Nice job!

I discussed this a bit with @evan2645 this morning. We had a similar conversation yesterday regarding a different PR. The conversation was about how strict we should be as validators of different parts of the spec. Evan's point was that we should probably only be very strict in validating things that have security implications or are likely the result of misuse or abuse. Other things, that are listed in the specification as required, might be so only for the sake of interoperability. Those kinds of things we can be a little less stringent about.

This function validates the chain and ensures that each certificate has:
1. proper SPIFFE ID shape (root v.s. non-root)
1. proper basic constraints (e.g. CA bit)
1. proper key usage (including CA bit)
1. proper extended key usage

Out of those three, the basic constraints and key usage seem pretty important in terms of security. The SPIFFE ID check and possibly the extended key usage, less so. This is why the `Verify` method currently only checks for basic constraints and key usage.

I think we probably want to make this a little more lenient, at a minimum with the removal of the SPIFFE ID validation, and possibly the extended key usage.

@evan2645 might have additional thoughts.

Reply:
No further thoughts from me, though maybe @marcosy has some feelings to the contrary?

Reply:
Ok! sounds good. I'll remove the SPIFFE ID and the extended key usage validations